widget/calendar.go

```diff
@@ -0,0 +1,223 @@
+package widget
+
+import (
+	"math"
+	"strconv"
+	"strings"
+	"time"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/lang"
+	"fyne.io/fyne/v2/layout"
+	"fyne.io/fyne/v2/theme"
+)
+
+// Declare conformity with Layout interface
+var _ fyne.Layout = (*calendarLayout)(nil)
+
+const (
+	daysPerWeek      = 7
+	maxWeeksPerMonth = 6
+)
+
+var minCellContent = NewLabel("22")
+
+// Calendar creates a new date time picker which returns a time object
+//
+// Since: 2.6
+type Calendar struct {
+	BaseWidget
+	currentTime time.Time
+
+	monthPrevious *Button
+	monthNext     *Button
+	monthLabel    *Label
+
+	dates *fyne.Container
+
+	OnChanged func(time.Time) `json:"-"`
+}
+
+// NewCalendar creates a calendar instance
+//
+// Since: 2.6
+func NewCalendar(cT time.Time, changed func(time.Time)) *Calendar {
+	c := &Calendar{
+		currentTime: cT,
+		OnChanged:   changed,
+	}
+
+	c.ExtendBaseWidget(c)
+
+	return c
+}
+
+// CreateRenderer returns a new WidgetRenderer for this widget.
+// This should not be called by regular code, it is used internally to render a widget.
+func (c *Calendar) CreateRenderer() fyne.WidgetRenderer {
+	c.monthPrevious = NewButtonWithIcon("", theme.NavigateBackIcon(), func() {
+		c.currentTime = c.currentTime.AddDate(0, -1, 0)
+		// Dates are 'normalised', forcing date to start from the start of the month ensures move from March to February
+		c.currentTime = time.Date(c.currentTime.Year(), c.currentTime.Month(), 1, 0, 0, 0, 0, c.currentTime.Location())
+		c.monthLabel.SetText(c.monthYear())
+		c.dates.Objects = c.calendarObjects()
+	})
+	c.monthPrevious.Importance = LowImportance
+
+	c.monthNext = NewButtonWithIcon("", theme.NavigateNextIcon(), func() {
+		c.currentTime = c.currentTime.AddDate(0, 1, 0)
+		c.monthLabel.SetText(c.monthYear())
+		c.dates.Objects = c.calendarObjects()
+	})
+	c.monthNext.Importance = LowImportance
+
+	c.monthLabel = NewLabel(c.monthYear())
+
+	nav := &fyne.Container{Layout: layout.NewBorderLayout(nil, nil, c.monthPrevious, c.monthNext),
+		Objects: []fyne.CanvasObject{c.monthPrevious, c.monthNext,
+			&fyne.Container{Layout: layout.NewCenterLayout(), Objects: []fyne.CanvasObject{c.monthLabel}}}}
+
+	c.dates = &fyne.Container{Layout: newCalendarLayout(), Objects: c.calendarObjects()}
+
+	dateContainer := &fyne.Container{Layout: layout.NewBorderLayout(nav, nil, nil, nil),
+		Objects: []fyne.CanvasObject{nav, c.dates}}
+
+	return NewSimpleRenderer(dateContainer)
+}
+
+func (c *Calendar) calendarObjects() []fyne.CanvasObject {
+	var columnHeadings []fyne.CanvasObject
+	for i := 0; i < daysPerWeek; i++ {
+		j := i + 1
+		if j == daysPerWeek {
+			j = 0
+		}
+
+		t := NewLabel(shortDayName(time.Weekday(j).String()))
```

Comment:
Does time.Weekday(j).String() always return the English names regardless of locale? Since this is just now being used as a localization key anyway, it may make sense to switch on the value of time.Weekday(j) itself, which may be more robust and easier to reason about

Reply:
Yes I looked at the source and that string doesn't seem to be localised.
If we use it directly then the key would be something like "day.1" is that clear enough do you think?

Reply:
I guess I was confused why we were still indexing the first 3 bytes of a string in shortDayName and thought this might help with that.

Reply:
I don't think it is indexing the first 3 - that is the fallback.
So if a language is missing the translation it uses the start of the word...

Reply:
Sorry, didn't realize lang.X was a 'with fallback' API - I'm used to using lang.L